### PR TITLE
FE: Monitoring Policies: Bug when selectedPolicy is undefined

### PR DESCRIPTION
**Pull request by faizanjanjua**
## Description
Description

In MonitoringPolicyBasicInformationEditAddForm and elsewhere, getting this error:

 

TypeError: Cannot read properties of undefined (reading 'name')
    at Proxy.<anonymous> (MonitoringPolicyBasicInformationEditAddForm-DU2WWxCH.js:1:824)
when trying to read store.selectedPolicy.name in v-model

Could be fixed by making sure monitoringPoliciesStore.selectedPolicy is always defined. But we will need a flag to note whether it's actually selected.

add isPolicySelected to store state

monitoringPolicyStore.clearSelectedPolicy should be as follows, and we need to update some cases where we check for store.selectedPolicy to instead check for store.isPolicySelected

clearSelectedPolicy() {
  this.selectedPolicy = cloneDeep(defaultPolicy)
  this.isPolicySelected = false
}
and and fix displayPolicyForm:

displayPolicyForm(policy?: Policy) {
  this.selectedPolicy = policy ? cloneDeep(policy) : cloneDeep(defaultPolicy)
  this.isPolicySelected = true
  this.selectedRule = undefined
}

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2469

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 1]
```diff
@@ -90,9 +90,9 @@ async function getDefaultRule(): Promise<PolicyRule> {
   }
 }
 
-export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore', {
-  state: (): TState => ({
-    selectedPolicy: undefined,
+export const  useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore', {
+  state: (): TState =>  ({
+    selectedPolicy: defaultPolicy,
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:95]: Change to `cloneDeep(defaultPolicy)`, we always want to make a clone of the `defaultPolicy`.

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 2]
```diff
@@ -108,6 +108,7 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
     async getMonitoringPolicies() {
       const queries = useMonitoringPoliciesQueries()
       await queries.listMonitoringPolicies()
+      this.selectedRule = await getDefaultRule()
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:111]: Add a `defaultRule` and then use `cloneDeep(defaultRule)` here.

Looks like we will probably get rid of `getDefaultRule()` and `getDefaultEventCondition()` eventually, they are only used in `Legacy` now.

```typescript
const defaultRule: PolicyRule = {
    id: 0,
    name: '',
    componentType: ManagedObjectType.Node,
    detectionMethod: DetectionMethod.Event,
    eventType: EventType.SnmpTrap,
    alertConditions: [],
    vendor: 'generic'
}

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 3]
```diff
@@ -121,12 +122,13 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
         p.enabled = true
       })
     },
-    displayPolicyForm(policy?: Policy) {
+    async displayPolicyForm(policy?: Policy) {
       this.selectedPolicy = policy ? cloneDeep(policy) : cloneDeep(defaultPolicy)
-      this.selectedRule = undefined
+      this.selectedRule = await getDefaultRule()
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:127]: Change to `cloneDeep(defaultRule)`

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 4]
```diff
@@ -121,12 +122,13 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
         p.enabled = true
       })
     },
-    displayPolicyForm(policy?: Policy) {
+    async displayPolicyForm(policy?: Policy) {
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:125]: Won't need `async` after below change.

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 5]
```diff
@@ -121,12 +122,13 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
...
       this.selectedPolicy = policy ? cloneDeep(policy) : cloneDeep(defaultPolicy)
-      this.selectedRule = undefined
+      this.selectedRule = await getDefaultRule()
     },
-    clearSelectedPolicy() {
-      this.selectedPolicy = undefined
+    async clearSelectedPolicy() {
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:129]: Remove `async`, change line below to `this.selectedRule = cloneDeep(defaultRule)`

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 6]
```diff
@@ -121,12 +122,13 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
...
+      this.selectedPolicy = defaultPolicy
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:130]: Should be `this.selectedPolicy = cloneDeep(defaultPolicy)`

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 7]
```diff
@@ -266,8 +268,8 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
       await addMonitoringPolicy({ policy })
 
       if (!error.value) {
-        this.selectedPolicy = undefined
-        this.selectedRule = undefined
+        this.selectedPolicy = defaultPolicy
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:271]: Change to `this.selectedPolicy = cloneDeep(defaultPolicy)` and `this.selectedRule = cloneDeep(defaultRule)`

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 8]
```diff
@@ -317,8 +319,8 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
       const { deleteMonitoringPolicy } = useMonitoringPoliciesMutations()
       await deleteMonitoringPolicy({ id: this.selectedPolicy?.id })
       this.getMonitoringPolicies()
-      this.selectedRule = undefined
-      this.selectedPolicy = undefined
+      this.selectedRule = await getDefaultRule()
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:322]: Same changes here.

File: ui/src/store/Views/monitoringPoliciesStore.ts [hunk 9]
```diff
@@ -327,17 +329,22 @@ export const useMonitoringPoliciesStore = defineStore('monitoringPoliciesStore',
...
+        await this.displayRuleForm(rule)
+        this.alertRuleDrawer = true
+      } else {
+        this.alertRuleDrawer = false
+      }
+
     },
     async closeAlertRuleDrawer() {
       this.alertRuleDrawer = false
-      this.selectedRule = undefined
```
Comment (synqotik) [inline ui/src/store/Views/monitoringPoliciesStore.ts:335]: Same changes as elsewhere

File: ui/tests/MonitoringPolicies/monitoringPolicies.test.ts [hunk 10]
```diff
@@ -78,7 +92,7 @@ describe.skip('Monitoring Policies', () => {
     const store = useMonitoringPoliciesStore()
     const newRuleBtn = wrapper.get('[data-test="new-rule-btn"]')
 
-    expect(store.selectedRule).toBeUndefined()
+    expect(store.selectedRule).toBeDefined()
```
Comment (synqotik) [inline ui/tests/MonitoringPolicies/monitoringPolicies.test.ts:95]: Do another check here that `id` is `0` and `name` is empty string, so we know it's the default rule.

File: ui/tests/MonitoringPoliciesLegacy/monitoringPoliciesLegacy.test.ts [hunk 11]
```diff
@@ -89,7 +101,7 @@ describe('Monitoring Policies Legacy', () => {
     const store = useMonitoringPoliciesStore()
     const newRuleBtn = wrapper.get('[data-test="new-rule-btn"]')
 
-    expect(store.selectedRule).toBeUndefined()
+    expect(store.selectedRule).toBeDefined()
```
Comment (synqotik) [inline ui/tests/MonitoringPoliciesLegacy/monitoringPoliciesLegacy.test.ts:104]: Same as other test.

**Review by synqotik — CHANGES_REQUESTED** [review]
See comments.

When resetting the policy or rule, need to make a clone/copy of the default objects, otherwise those values will get changed.

Also I think we can just a simple object for `defaultRule`. The code from before makes API calls which is expensive and I don't it's needed. We will eventually remove that code when Legacy components are deleted.

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 3 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 2 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved , thanks

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 1 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved , thanks

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 5 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved , thanks

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 7 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved , thanks

File: ui/src/store/Views/monitoringPoliciesStore.ts (hunk 8 above)
Comment (faizanjanjua) [reply ui/src/store/Views/monitoringPoliciesStore.ts]: Resolved , thanks

File: ui/tests/MonitoringPolicies/monitoringPolicies.test.ts (hunk 10 above)
Comment (faizanjanjua) [reply ui/tests/MonitoringPolicies/monitoringPolicies.test.ts]: Resolved , thanks

File: ui/tests/MonitoringPoliciesLegacy/monitoringPoliciesLegacy.test.ts (hunk 11 above)
Comment (faizanjanjua) [reply ui/tests/MonitoringPoliciesLegacy/monitoringPoliciesLegacy.test.ts]: Resolved , thanks

**Review by synqotik — APPROVED** [review]
LGTM

**Pull request merged**